---
layout: vanilla
permalink: /docs/
description: "Tutorials for Mouse Atlas."
modified: 2018-01-13
tags: [mouse, manual, vignette, atlas]
---
<!DOCTYPE html>
<html lang= "{{ page.lang | default: site.lang | default: "en" }}">

  {% include head.html %}

  <!-- See here for how to use the TOC control:
  https://afeld.github.io/bootstrap-toc/ -->
  <body  data-spy= "scroll" data-target= "#toc">
    {% include header.html %}

    <div class= "container">
      <div class= "row">
        <!-- sidebar, which will move to the top on a small screen -->
        <div class= "col-sm-4 sidenav">
          <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
        </div>
        <!-- main content area -->
        <div class= "col-sm-8">

    <h2>Loading Matrix Files</h2>
    <p>We provide two major formats for all matrix data, matrix-market and RDS.</p>
    <h3>Matrix Market Format</h3>
    <p>Matrix market format is the format used by the <code>cellranger</code> pipeline from 10X genomics, which may be familiar to many of you. This format is simply a text file that allows reconstruction of a sparse matrix, along with the peak (or gene) and cell names that specify the row and column names of the matrix, respectively.</p>

    <p>This format is readable in <code>R</code>:
    <figure class="highlight">
    <pre><code class="language-r" data-lang="r">library(Matrix)
my_matrix = readMM("my_matrix.mtx.gz")
row.names(my_matrix) = read.delim('my_matrix.peaks.txt', header=FALSE)$V1
colnames(my_matrix) = read.delim('my_matrix.cells.txt', header=FALSE)$V1</code></pre>
    </figure>

<div class="spacer"></div>

    <p>as well as <code>python</code> (based on code from <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices">10X Genomics Documentation</a>):</p>

    <figure class="highlight">
    <pre><code class="language-python" data-lang="python">import csv
import scipy.io

mat = scipy.io.mmread("my_matrix.mtx.gz")
peaks = [row[0] for row in csv.reader(open("my_matrix.peaks.txt"), delimiter="\t")]
cells = [row[0] for row in csv.reader(open("my_matrix.cells.txt"), delimiter="\t")]</code></pre>
    </figure>

<div class="spacer"></div>

<p>In principle this text format is readable in many other languages as well.</p>

    <h3>RDS Format</h3>
    <p>We also provide these matrices in RDS format, which is easily readable in <code>R</code>, but has less cross-compatability with other languages. Row and column names are already associated with the matrix.</code></p>

    <figure class="highlight">
    <pre><code class="language-r" data-lang="r">library(Matrix)
my_matrix = readRDS("my_matrix.rds")</code></pre>
    </figure>


</div>
</div>
</div>

<br>
<br>
<br>
<br>

{% include footer.html %}

</body>

</html>

